chai/cmd_executor.py

```diff
@@ -0,0 +1,225 @@
+            kinds are as follows
+
+            - Error: A checking violation is found.
+            - Deadlock: A deadlock was found.
+            - RuntimeError: A runtime error was encountered, preventing checking.
+        counter_examples: A list of counterexamples found.
+    """
+
+    checking_result: str
+    counter_example: List[Counterexample]
+
+    @classmethod
+    def _of_dict(cls, d: dict) -> ParsingError | TypecheckingError | CheckingError:
+        _ensure_is_pass_failure(d)
+
+        error_data = d["data"]["error_data"]
+        pass_name = d["data"]["pass_name"]
+        if pass_name == "BoundedChecker":
+            checking_result = error_data["checking_result"]
+            if checking_result == "Deadlock":
+                # TODO We should use the same key for both counterexamples
+                counter_examples = error_data["counterexample"]
+            else:
+                counter_examples = error_data["counterexamples"]
+            # TODO Handle all other checking errors
+            return CheckingError(pass_name, checking_result, counter_examples)
+        else:
+            return TypecheckingError._of_dict(d)
+
+
+E = TypeVar("E")
+CmdExecutorResult = RpcResult[E | TlaModule]
+
+
+def _str_source(spec, aux):
+    return {"source": {"type": "string", "content": spec, "aux": aux}}
+
+
+def _config_json(spec: Input, aux: Optional[List[Input]], cfg: Optional[dict]) -> str:
+    # TODO: error if source already set in config?
+    aux = aux or []
+    config = cfg or {}
+    return json.dumps(config | {"input": _str_source(spec, aux)})
+
+
+class ChaiCmdExecutor(client.Chai[service.CmdExecutorStub]):
+    r"""Client for Shai's `CmdExecutor` service
+
+    The `CmdExecutor` service is a stateless service exposing the functionality
+    of Apalache's CLI.
+
+    Example usage:
+
+    ```
+    from chai import ChaiCmdExecutor
+
+    async with ChaiCmdExecutor.create() as client:
+        assert client.is_connected()
+        spec = '''
+            ---- MODULE M ----
+            VARIABLES
+                \* @type: Bool;
+                x,
+                \* @type: Bool;
+                y
+
+            Init == x = TRUE /\ y = TRUE
+            Next == x' = FALSE /\ y' = y
+            Inv == x
+            ====
+            '''
+        res = await client.check(spec, config={"checker": {"inv": ["Inv"]}})
+        assert isinstance(res, CheckingError)
+        assert res.checking_result == "Error"
+        states = res.counter_example[0]["states"][1]
+        assert states == {"#meta": {"index": 1}, "x": False, "y": True}
+
+    See the documetation of :class:`~chai.client.Chai` for instruction on using
+    the client safely without a context manager.
+    ```
+    """
+
+    PING_REQUEST = msg.PingRequest  # type: ignore
+
+    @classmethod
+    def _service(cls, channel: aio.Channel) -> service.CmdExecutorStub:
+        return service.CmdExecutorStub(channel)
```

Comment:
Can't find where this is used?

Reply:
This is required by the `client.Chai` ABC.